arena.h

```diff
@@ -95,7 +95,61 @@ void free_region(Region *r)
 #elif ARENA_BACKEND == ARENA_BACKEND_LINUX_MMAP
 #  error "TODO: Linux mmap backend is not implemented yet"
 #elif ARENA_BACKEND == ARENA_BACKEND_WIN32_VIRTUALALLOC
-#  error "TODO: Win32 VirtualAlloc backend is not implemented yet"
+
+#if !defined(_WIN32)
+#  error "Current platform is not Windows"
+#endif
+
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+
+#define INV_HANDLE(x)       (((x) == NULL) || ((x) == INVALID_HANDLE_VALUE))
+
+Region *new_region(size_t capacity)
+{
+    SIZE_T size_bytes = sizeof(Region) + sizeof(uintptr_t) * capacity;
+    Region *r = VirtualAllocEx(
+        GetCurrentProcess(),      /* Allocate in current process address space */
+        NULL,                     /* Unknown position */
+        size_bytes,               /* Bytes to allocate */
+        MEM_COMMIT | MEM_RESERVE, /* Reserve and commit allocated page */
+        PAGE_READWRITE            /* Permissions ( Read/Write )*/
+    );
+    if (INV_HANDLE(r))
+        ARENA_ASSERT(0 && "VirtualAllocEx() failed.");
+
+    r->next = NULL;
+    r->count = 0;
+    r->capacity = capacity;
+    return r;
+}
+
+void free_region(Region *r)
+{
+    if (INV_HANDLE(r))
+        return;
+
+    BOOL free_result = VirtualFreeEx(
+        GetCurrentProcess(),        /* Deallocate from current process address space */
+        (LPVOID)r,                  /* Address to deallocate */
+        0,                          /* Bytes to deallocate ( Unknown, deallocate entire page ) */
+        MEM_DECOMMIT                /* Decommit page */
+    );
+
+    if (FALSE == free_result)
+        ARENA_ASSERT(0 && "VirtualFreeEx() failed.");
+
+    free_result = VirtualFreeEx(
+        GetCurrentProcess(),        /* Deallocate from current process address space */
+        (LPVOID)r,                  /* Address to deallocate */
+        0,                          /* Bytes to deallocate ( Unknown, deallocate entire page ) */
+        MEM_RELEASE                 /* Release page */
+    );
```

Comment:
I'm not an expert on WinAPI, but do you really have to decommit and release as two separate actions? `dwFreeType` seems like a bit mask. Is it possible to combine both actions `MEM_DECOMMIT | MEM_RELEASE`?

Reply:
Unfortunatelly, no, it's not. 
When I tried to pass both ```MEM_DECOMMIT | MEM_RELEASE```, it failed:

```
test.c: ARENA_BACKEND_WIN32_VIRTUALALLOC works!
Assertion failed: 0 && "VirtualFreeEx() failed.", file arena.h, line 140 
```

After searching and digging in MSDN, I found out that functions ```VirtualAllocXX``` [will always crash with parameters ```MEM_DECOMMIT | MEM_RELEASE```](https://github.com/MicrosoftDocs/cpp-docs/blob/main/docs/code-quality/c6331.md).

From **Warning C6331**:
> Therefore, the values MEM_DECOMMIT and MEM_RELEASE may not be used together in the same call.

> It's not required for decommit and release to occur as independent steps. Releasing committed memory will decommit the pages as well.

So [I left only one call with ```MEM_RELEASE``` argument](https://github.com/Nikita-bunikido/arena/commit/58b79eeca9e3bc02570fecaf092c7717d7ba1f37), which will implicitly decommit the page.